### PR TITLE
Support KernelSU / Apatch & fix broken Frida server launch

### PR DESCRIPTION
**Pull request by enovella**
Screenshot 2024-01-27 at 23 15 36

Screenshot 2024-01-27 at 23 08 46

### CODE REVIEW

**Comment by ViRb3** [comment]
@enovella Why does the auto restart loop not work for you? I see no reason why it wouldn't work.

**Comment by ViRb3** [comment]
@enovella Sorry, meant to test myself but got side tracked. I just pulled out my OnePlus 5T running Elixir OS 2.1 (Android 12L), Magisk 27.0, MagiskFrida 16.2.1. It works perfectly as I can debug applications from my computer over USB, and I can also see the process inside a shell in the phone:
```bash
$ adb shell
OnePlus5T:/ $ pgrep -l frida
3938 frida-server
```

What exact setup did you test across your devices? I honestly find it very strange for the daemonization to work and regular blocking mode not.

**Comment by liangjaden** [comment]
这个PR会继续跟进不？期待能支持kernelsu

**Comment by sha3rawi33** [comment]
bump

**Comment by nEdAy** [comment]
looking forward to it supporting KernelSU.

**Comment by ViRb3** [comment]
Can all of you please let me know whether the current MagiskFrida with auto restart loop works for you, and you are able to debug apps using Frida on your PC? I am happy to merge this PR but I want to figure out the problem @enovella is experiencing.

**Comment by ViRb3** [comment]
Yeah, I know it doesn't work for you, but it works for me this is the confusing part. So I am asking other people, apart from you and me, to test as well and we can get a better picture. I don't want to make changes without fully understanding them.

**Comment by liangjaden** [comment]
Yeah，this pr is work for me！Nokia 6.1plus，pixel OS



---Original---
From: ***@***.***&gt;
Date: Fri, May 24, 2024 21:03 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ViRb3/magisk-frida] Support KernelSU &amp; fix broken Frida serverlaunch (PR #46)




 
Yeah, I know it doesn't work for you, but it works for me this is the confusing part. So I am asking other people, apart from you and me, to test as well and we can get a better picture. I don't want to make changes without fully understanding them.

**Comment by ViRb3** [comment]
@liangjaden can you test the latest official magisk-frida release, not this PR? https://github.com/ViRb3/magisk-frida/releases/download/16.2.5-1/MagiskFrida-16.2.5-1.zip

**Comment by liangjaden** [comment]
Maybe next week, because the phone is
In the company




---Original---
From: ***@***.***&gt;
Date: Fri, May 24, 2024 21:12 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ViRb3/magisk-frida] Support KernelSU &amp; fix broken Frida serverlaunch (PR #46)




 
@liangjaden can you test the latest official magisk-frida release, not this PR? Link: https://github.com/ViRb3/magisk-frida/releases/download/16.2.5-1/MagiskFrida-16.2.5-1.zip

**Comment by ys1231** [comment]
@enovella His problem can be solved by ending the server process once, and the loop can also work normally

**Comment by liangjaden** [comment]
@ViRb3 After testing, the latest version is working on my phone

**Comment by ViRb3** [comment]
@enovel unfortunately I only have one device at hand. The stable version of this package has used the loop for over a year and you're the first person to say it doesn't work, so I am just very confused why. I really want to get to the bottom of this before making any changes. I will try to run a different ROM over the next few days, also I'll also try an emulator.

For anyone reading this ticket, please also test and let me know.

**Comment by 0xE0-rng** [comment]
@enovella I tried your module but does not seem to work on my Pixel 7 (Android 14, Kernel Patch 0.10.7, Apatch 10763)
I am still debugging, but one first issue seems to be that $KSU does not seems to be set which leads to the module not recognizing that its a KernelSU install. 
I cannot attach screenshots, but the install does seem to work except the $KSU var not being set. 

Following that the extraction just seems to fail for me... The zip file contains the frida server
```
~/projects/magisk-frida-edu/build  ksufrida ± > unzip -l MagiskFrida-16.3.3-1.zip
Archive:  MagiskFrida-16.3.3-1.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
      215  07-02-2024 12:30   module.prop
     7325  07-02-2024 11:44   install.sh
      557  06-28-2024 17:11   README.md
        8  06-28-2024 17:11   META-INF/com/google/android/updater-script
     3794  06-28-2024 17:11   META-INF/com/google/android/update-binary
      318  06-28-2024 17:11   common/post-fs-data.sh
       84  06-28-2024 17:11   common/system.prop
      503  07-02-2024 11:44   common/service.sh
 56466624  07-02-2024 12:30   files/frida-server-arm64
 56532312  07-02-2024 12:30   files/frida-server-x86
113371768  07-02-2024 12:30   files/frida-server-x86_64
 27129236  07-02-2024 12:30   files/frida-server-arm
---------                     -------
253512744                     12 files
```

but its not extracted into the correct folder:

```
 ~ adb logcat | grep -i frida
- waiting for device -
07-02 15:32:34.246   864   864 I APatchD : apd::mount: mount overlayfs on "/system", lowerdir=/data/adb/modules/magisk-frida/system:/system, upperdir=None, workdir=None
```

```
panther:/data/adb/modules/magisk-frida # ls -lah system/bin
total 4
drwxr-xr-x 2 root root 1.0K 2024-07-02 15:24 .
drwxr-xr-x 3 root root 1.0K 2024-07-02 15:24 ..
panther:/data/adb/modules/magisk-frida # ls /system/bin/fr*
/system/bin/free
```


Apologies if thats an obious error, by magisk / Ksu expiriece is limited.... Let me know if I can help somehow

**Comment by 0xE0-rng** [comment]
Ah, that makes alot of sense, I just read the docs on APatch and ajusted your script and it all works. I will create a PR later today to your repo :)

**Comment by 0xE0-rng** [comment]
https://github.com/enovella/magisk-frida-1/pull/1

File: base/install.sh [hunk 1]
```diff
@@ -141,14 +141,33 @@ on_install() {
...
 
-  F_TARGETDIR="$MODPATH/system/bin"
-  UNZIP="/data/adb/magisk/busybox unzip"
+  if [ "$BOOTMODE" ] && [ "$KSU" ]; then
+      ui_print "- Installing from KernelSU app"
+      ui_print "- KernelSU version: $KSU_KERNEL_VER_CODE (kernel) + $KSU_VER_CODE (ksud)"
+      UNZIP="/data/adb/ksu/bin/busybox unzip"
+  elif [ "$BOOTMODE" ] && [ "$APATCH" ]; then
+      ui_print "- Installing from APatch app"
+      ui_print "- APatch version: $APATCH_VER_CODE. Magisk version: $MAGISK_VER_CODE"
```
[the remark this answers is not in the document]
Comment (0xE0-rng) [reply base/install.sh:151]: It's available (as seen on the output above). Thanks for the feedback, I will give it a look when I have some free time :)

**Comment by ViRb3** [comment]
Sorry @enovella, I didn't mean to drag this for so long, just thought I'd have free time to dig into it but still haven't. I'd be happy to archive this repo in favor of yours, but it has a build hook set up with Frida and apparently very good Google SEO, so it would be a pain to migrate.

Let's do this - from now on I will merge your PRs with just a code review, and if there's any issues, we can always revert.

Apologies again for dragging this so long, let's hope this new approach works out for the best!

**Pull request merged**